### PR TITLE
Unify durations 

### PR DESCRIPTION
**Pull request by Nathan-Fenner**
When using relative dates with `parseDate` (which now asks for a `now` time), the format is identical to `parseDuration`:

The units supported are `s`, `m`, `h`, `hr`, `d`, `w` (week = 7 days), `M`, `mo`, (month = 30 days), `y`, `yr`, (year = 365 days).

### CODE REVIEW

File: query/parser.go [hunk 1]
```diff
@@ -99,57 +99,55 @@ var dateFormats = []string{
...
 		}
-
-		offset := offsetSeconds * 1000
-		specifier := matches[2]
-
-		switch specifier {
+		switch matches[3] {
 		case "s":
-			// Already in millis, do nothing.
+			offset *= 1
```
Comment (achow) [inline query/parser.go]: This does nothing. Should just leave it as-is.

File: query/parser.go [hunk 2]
```diff
@@ -99,57 +99,55 @@ var dateFormats = []string{
...
+		case "w":
+			offset *= 60 * 60 * 24 * 7
 		case "M":
 			offset *= 60 * 60 * 24 * 30
 		case "y":
 			offset *= 60 * 60 * 24 * 365
 		default:
-			return -1, fmt.Errorf("invalid relative timestamp %s", s)
+			// Won't happen, regex can't produce any other value.
+			panic(matches[3])
```
Comment (achow) [inline query/parser.go]: error, don't panic. This is recoverable, especially if we're using a REPL.

Comment (Nathan-Fenner) [inline query/parser.go]: The code is unreachable, provided the timestamp regexp (and the switch) is correct/consistent, so I decided to panic.

**Comment by achow** [comment]
Why require 'now - XX'? This is meant to be syntactic sugar, otherwise it just adds more friction when you're pulling metrics from the UI or otherwise building queries.

**Comment by Nathan-Fenner** [comment]
TODO: allow `'-10h'` as a valid relative timestamp

(originally left out due to (erroneous) concerns regarding conflicts between `parseDuration` and `parseDate`, now resolved)

**Comment by Nathan-Fenner** [comment]
Now short relative timestamps (`'-10h'`) are supported again.

File: query/parser.go [hunk 3]
```diff
@@ -99,57 +99,62 @@ var dateFormats = []string{
...
+	relativeTimeRegexpShort = regexp.MustCompile(`(-)\s*([0-9]+)([smhdwMy])$`)
+)
 
-// parseRelativeTime transforms a relative time string (e.g. `-10m`) to a
-// timestamp in millis.
-func parseRelativeTime(s string, now time.Time) (int64, error) {
-	if strings.ToLower(s) == "now" {
-		return now.Unix() * 1000, nil
+// parseDate converts the given datestring (from one of the allowable formats) into a millisecond offset from the Unix epoch.
+func parseDate(date string) (int64, error) {
```
Comment (achow) [inline query/parser.go]: Why remove `now`? Do tests even pass?

**Comment by achow** [comment]
If we support both `now - 10h` and `-10h`, what are the benefits of having the former?

**Comment by Nathan-Fenner** [comment]
There isn't much advantage. I'll pull the longer way out, and unify with duration parsing a bit more.

File: query/parser.go [hunk 4]
```diff
@@ -99,57 +99,60 @@ var dateFormats = []string{
...
 
-// `now` isn't captured in this regex
-var relativeTimeRegexp = regexp.MustCompile(`^-(\d+)([a-zA-Z])$`)
+var (
+	relativeTimeRegexp      = regexp.MustCompile(`^now\s*([+-])\s*([0-9]+)([smhdwMy])$`)
+	relativeTimeRegexpShort = regexp.MustCompile(`(-)\s*([0-9]+)([smhdwMy])$`)
+)
+
+// parseDate converts the given datestring (from one of the allowable formats) into a millisecond offset from the Unix epoch.
+func parseDate(date string, now time.Time) (int64, error) {
```
Comment (achow) [inline query/parser.go]: If you keep the relative time conversion as a separate function, you can reuse it for parsing the `resolution` duration.

File: query/value.go [hunk 5]
```diff
@@ -88,6 +89,8 @@ func (value scalarValue) toScalar() (float64, error) {
 	return float64(value), nil
 }
 
+var durationRegexp = regexp.MustCompile(`^([+-]?[0-9]+)([smhdwMy])$`)
```
Comment (achow) [inline query/value.go]: Can probably do without the `+` since omission is implicitly positive.

File: query/parser.go [hunk 6]
```diff
@@ -99,57 +99,16 @@ var dateFormats = []string{
...
-		case "y":
-			offset *= 60 * 60 * 24 * 365
-		default:
-			return -1, fmt.Errorf("invalid relative timestamp %s", s)
-		}
+// parseDate converts the given datestring (from one of the allowable formats) into a millisecond offset from the Unix epoch.
+func parseDate(date string, now time.Time) (int64, error) {
 
-		return now.Unix()*1000 - offset, nil
+	relativeTime, err := toDuration(stringValue(date))
```
Comment (achow) [inline query/parser.go]: Should ensure this duration isn't positive, since queries into the future don't really make sense.

File: query/value.go [hunk 7]
```diff
@@ -96,9 +99,30 @@ func toDuration(value value) (int64, error) {
 	if err != nil {
 		return 0, err
 	}
-	duration, err := time.ParseDuration(timeString)
+	matches := durationRegexp.FindStringSubmatch(timeString)
+	if matches == nil {
+		return 0, fmt.Errorf("expected duration to be of the form `[0-9]+[smhdwMy]`")
+	}
+	duration, err := strconv.ParseInt(matches[1], 10, 0)
```
Comment (achow) [inline query/value.go]: Needs to handle negatives.

File: query/parser.go [hunk 8]
```diff
@@ -402,9 +364,10 @@ func (p *Parser) insertPropertyKeyValue() {
 	case "to":
 		var unix int64
 		var err error
-		if unix, err = parseDate(value); err == nil {
+		now := time.Now()
+		if unix, err = parseDate(value, now); err == nil {
```
Comment (achow) [inline query/parser.go]: This should turn into a single `err != nil` since there's no second `if` anymore.

File: query/value.go (hunk 7 above)
Comment (Nathan-Fenner) [inline query/value.go]: Handle in what way?

File: query/parser.go [hunk 9]
```diff
@@ -418,13 +381,19 @@ func (p *Parser) insertPropertyKeyValue() {
 	case "resolution":
 		// The value must be determined to be an int if the key is "resolution".
 		intValue, err := strconv.ParseInt(value, 10, 64)
-		if err != nil {
-			p.flagSyntaxError(SyntaxError{
-				token:   value,
-				message: fmt.Sprintf("Expected number but parse failed; %s", err.Error()),
-			})
+		if err == nil {
```
Comment (achow) [inline query/parser.go]: See if you can turn this into a chain of `if`s instead of `break`s.

File: query/value.go [hunk 10]
```diff
@@ -88,17 +89,42 @@ func (value scalarValue) toScalar() (float64, error) {
...
+	case "h":
+		scale = 1000 * 60 * 60
+	case "d":
+		scale = 1000 * 60 * 60 * 24
+	case "w":
+		scale = 1000 * 60 * 60 * 24 * 7
+	case "M":
+		scale = 1000 * 60 * 60 * 24 * 30
+	case "y":
+		scale = 1000 * 60860 * 24 * 365
```
Comment (achow) [inline query/value.go]: Where does 60860 come from?

Comment (Nathan-Fenner) [inline query/value.go]: Typo, and good catch. 8 is asterisk key.

File: main/query.go [hunk 11]
```diff
@@ -37,7 +37,7 @@ func main() {
 	}
 
 	apiInstance := common.NewAPI()
-	myBackend := blueflood.NewBlueflood(*common.BluefloodUrl, *common.BluefloodTenantId)
+	myBackend := blueflood.NewBlueflood(blueflood.BluefloodClientConfig{*common.BluefloodUrl, *common.BluefloodTenantId})
```
Comment (jeeyoungk) [inline main/query.go]: this _shouldn't_ be needed if we rebase.

Comment (Nathan-Fenner) [inline main/query.go]: It looks like `main/query.go` got skipped over somehow, so it's not currently in master.

File: query/value.go [hunk 12]
```diff
@@ -88,17 +89,42 @@ func (value scalarValue) toScalar() (float64, error) {
 	return float64(value), nil
 }
 
+var durationRegexp = regexp.MustCompile(`^([+-]?[0-9]+)([smhdwMy]|ms)$`)
```
Comment (jeeyoungk) [inline query/value.go]: this is slowly getting more complex :P

File: main/query.go (hunk 11 above)
Comment (jeeyoungk) [inline main/query.go]: ahh :/

File: query/parser_test.go [hunk 13]
```diff
@@ -35,16 +35,20 @@ func Test_parseRelativeTime(t *testing.T) {
 		{"-3m", 1413321686000, true},
 		{"-4h", 1413307466000, true},
 		{"-5d", 1412889866000, true},
+		{"-3w", 1411507466000, true},
```
Comment (achow) [inline query/parser_test.go]: Needs a test for `ms`

File: query/value.go [hunk 14]
```diff
@@ -88,17 +89,42 @@ func (value scalarValue) toScalar() (float64, error) {
...
+		scale = 1000 * 60 * 60
+	case "d":
+		scale = 1000 * 60 * 60 * 24
+	case "w":
+		scale = 1000 * 60 * 60 * 24 * 7
+	case "M":
+		scale = 1000 * 60 * 60 * 24 * 30
+	case "y":
+		scale = 1000 * 60 * 60 * 24 * 365
 	}
```
Comment (achow) [inline query/value.go]: Needs a default case.

File: query/value.go [hunk 15]
```diff
@@ -88,17 +89,42 @@ func (value scalarValue) toScalar() (float64, error) {
...
 // It converts the return value to milliseconds.
 func toDuration(value value) (int64, error) {
 	timeString, err := value.toString()
 	if err != nil {
-		return 0, err
+		return -1, err
+	}
+	matches := durationRegexp.FindStringSubmatch(timeString)
+	if matches == nil {
+		return -1, fmt.Errorf("expected duration to be of the form `[0-9]+[smhdwMy]`")
```
Comment (achow) [inline query/value.go]: You can `regexp.String()` to pull the regex out of the `Regexp`

File: query/parser_test.go (hunk 13 above)
Comment (Nathan-Fenner) [inline query/parser_test.go]: There's now exhaustive tests for suffixes in `value_test.go`

**Comment by jeeyoungk** [comment]
LGTM on my side - @achow ?

**Pull request merged**